src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs

```diff
@@ -530,7 +531,9 @@ public static bool IsUnnecessaryCast(this CastExpressionSyntax cast, SemanticMod
 
                 if (castToOuterType.IsIdentity &&
                     !expressionToCastType.IsUnboxing &&
-                    expressionToCastType == expressionToOuterType)
+                    expressionToCastType == expressionToOuterType &&
+                    // if the conversion to the outer type doesn't exist, then we shouldn't offer, except for anonymous functions which can't be reasoned about
+                    (expressionToOuterType.Exists || originalIsAnonymousFunction))
```

Comment:
Here I'm left wondering what "original" is -- is that `expressionToCastType` or `expressionToOuterType`. I'd rather just see the access to `expressionToOuterType.IsAnonymousFunction` here. That's already very clear and consistent with the rest of the method.

Reply:
The problem is that `expressionToOuterType` is reassigned on [line 461](https://github.com/dotnet/roslyn/pull/27394/files/c20a8229418525a5a9b7d49bcc0fb43ecbdfd6b1#diff-6f8f267309a989ee1a5a92a703afe42bR461) so I needed to capture the state of the previous value before that reassignment.

Calling `GetSpeculatedExpressionToOuterTypeConversion` again would be too costly I'm assuming. Would a rename of the local help? I could try putting in an earlier exit for this case and see if it breaks any other tests.

Reply:
Ah! That makes sense. In that case, there are a couple of options:

1. Rename the local to better describe what is expected in this condition. This might get long. E.g. `originallySpeculatedExpressionToOuterTypeWasAnonymousFunctionConversion` -- though I'm not sure that's not quite right. :smile:
2. Change the logic that reassigns `expressionToOuterType` to capture the original conversion. I could image something like this:

```C#
var speculativeExpressionToOuterType = GetSpeculatedExpressionToOuterTypeConversion(speculationAnalyzer.ReplacedExpression, speculationAnalyzer, cancellationToken);

// CONSIDER: Anonymous function conversions cannot be compared from different semantic models as lambda symbol comparison requires syntax tree equality. Should this be a compiler bug?
// For now, just revert back to computing expressionToOuterType using the original semantic model.
var expressionToOuterType = speculativeExpressionToOuterType.IsAnonymousFunction
    ? semanticModel.ClassifyConversion(cast.SpanStart, cast.Expression, outerType)
    : speculativeExpressionToOuterType;
```

Then, you could test `speculativeExpressionToOuterType.IsAnonymousFunction` here.